elastic/eui
# [EuiDataGrid] Display selector / toolbar control enhancements

## Summary

This PR adds 3 separate enhancements (listed with screenshots below). As always, I strongly recommend [following along by commit](https://github.com/elastic/eui/pull/8080/commits), as I tried to split out several code refactors out to make following the actual new features easier.

---

### Redesigned row height form controls (https://github.com/elastic/eui/pull/8080/commits/493e2e0fa4820f0e391fbc13b95b866726db2341)
Per @MichaelMarcialis's designs in https://github.com/elastic/eui/issues/7835#issuecomment-2197604786 (although these changes do not close the issue, 

| Single/Default | Custom `lineCount` | Auto fit |
|--------|--------|--------|
| <img width="493" alt="" src="https://github.com/user-attachments/assets/aba8f7aa-4b8e-4822-9e95-660b82bac9d0"> | <img width="641" alt="" src="https://github.com/user-attachments/assets/8a4b8790-ad63-4fa7-95ee-0fa1184d5f21"> | <img width="636" alt="" src="https://github.com/user-attachments/assets/0f5b2300-878d-4cfc-b50e-235e0dee681b"> |

---

### Completely custom control over rendering the display selector (https://github.com/elastic/eui/pull/8080/commits/92bc73e7d52377bf5c6819348a61acc8359b4304, https://github.com/elastic/eui/pull/8080/commits/81b5e25d3e2ac16e02969696f81bdba84018fa3d) (closes #7951)

Consumers can now pass a `toolbarVisibility.showDisplaySelector.customRender` render prop/function that passes back the default density/row height/reset button as optional nodes to render. This allows consumers to reorder and otherwise completely customize the display selector.

<img width="500" alt="" src="https://github.com/user-attachments/assets/c20b0098-8d68-48de-9065-042c2a12d86d">

```tsx
const customDisplaySelectorRender = ({ densityControl }) => (
  <div>Custom display selector {densityControl}</div>
);

<EuiDataGrid
  // ... required props
  toolbarVisibility={{
    showDisplaySelector: {
      customRender: customDisplaySelectorRender
    }
  }}
/>
```

---

### `gridStyle` updates after users have changed grid styles via the display selector (https://github.com/elastic/eui/pull/8080/commits/9cf4a4da1ff44c293053cb4115ccaade6820ee5b) (closes #7962)

Previously, EuiDataGrid would always defer to the user's row height/density selection from the display selector controls, and if consumers tried to pass in new `gridStyle` updates after, or control them from an external button, those updates would be ignored.

You can reproduce this [on production](https://eui.elastic.co/storybook/?path=/story/tabular-content-euidatagrid-gridstyle-prop--playground) by changing the density in the toolbar display selector and then trying to change `fontSize` or `cellPadding` in the Storybook controls - nothing happens at that point.

This PR changes this behavior so that new `gridStyle` updates are reflected in the datagrid (see QA steps below).

## QA

- New row height UI should be testable on any datagrid:
  - https://eui.elastic.co/pr_8080/#/tabular-content/data-grid
- Custom render testing:
  - https://eui.elastic.co/pr_8080/#/tabular-content/data-grid-toolbar#completely-custom-toolbar-rendering
- `gridStyles` updates should be testable on Storybook:
  - Go to https://eui.elastic.co/pr_8080/storybook/?path=/story/tabular-content-euidatagrid-gridstyle-prop--playground&args=fontSize:s
  - Go to the display selector in the toolbar and change, e.g. density to `Expanded`
  - [x] In the Storybook controls, change `fontSize` to `s` and confirm that the datagrid correctly updates to this setting

### General checklist

- Browser QA
    ~- [ ] Checked in both **light and dark** modes~
    ~- [ ] Checked in **mobile**~
- Docs site QA
    ~- [ ] Added **[documentation](https://github.com/elastic/eui/blob/main/wiki/contributing-to-eui/documenting)**~
    ~- [ ] Checked **[Code Sandbox](https://codesandbox.io/)** works for any docs examples~
- Code quality checklist
- Release checklist
    ~- [ ] If applicable, added the **breaking change** issue label (and filled out the breaking change checklist)~
- Designer checklist
